Value Checker ignores annotations on casts

I realize this is not very important because users would rarely write annotations on casts, but it is quite unexpected behavior.

```java
import org.checkerframework.common.value.qual.IntVal;
public class CastReport {
 void intcasts(@IntVal(1) int i){
  @IntVal(1) int z1 = (@IntVal(2) int)i;
  @IntVal(2) int z2 = (@IntVal(2) int)i;
 }
}
```
The Value checker ignores the annotation on the cast and keeps the qualifier of the operand, giving this warning:
```
CastReport.java:5: error: [assignment.type.incompatible] incompatible types in assignment.
		@IntVal(2) int z2 = (@IntVal(2) int)i;
		                    ^
  found   : @IntVal(1) int
  required: @IntVal(2L) int
1 error
```

I think one of the following ways would be better than the current situation:

1. Use the annotation on the cast and ignore the qualifier of the operand. Warn about unsafe cast if the qualifier on the cast is not a subtype of the qualifier of the operand.
2. Use the greatest lower bound of the qualifiers from the cast and the operand. This would give the best precision. Again, warn about unsafe cast if the qualifier on the cast is not a subtype of the qualifier of the operand.

(I am interested in what is the best way to handle that because I need to modify the implementation of casting in the value checker to correctly handle `@ArrayLen` on strings, and I would rather not implement the wrong logic.)

--- a/framework/src/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java
+++ b/framework/src/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java
@@ -945,11 +945,11 @@ public class ValueAnnotatedTypeFactory extends BaseAnnotatedTypeFactory {
                         List<?> values = ValueCheckerUtils.getValuesCastedToType(oldAnno, newType);
                         newAnno = createResultingAnnotation(atm.getUnderlyingType(), values);
                     }
-                    atm.replaceAnnotation(newAnno);
+                    atm.addMissingAnnotations(Collections.singleton(newAnno));
                 }
             } else if (atm.getKind() == TypeKind.ARRAY) {
                 if (tree.getExpression().getKind() == Kind.NULL_LITERAL) {
-                    atm.replaceAnnotation(BOTTOMVAL);
+                    atm.addMissingAnnotations(Collections.singleton(BOTTOMVAL));
                 }
             }
             return null;

--- a/framework/tests/value/Basics.java
+++ b/framework/tests/value/Basics.java
@@ -163,6 +163,13 @@ class Basics {
         @IntRange(from = 0, to = 30) int test9 = a;
     }
 
+    public void intCastTest(@IntVal({0, 1}) int input) {
+        @IntVal({0, 1}) int c = (int) input;
+        @IntVal({0, 1}) int ac = (@IntVal({0, 1}) int) input;
+        //:: warning: (cast.unsafe)
+        @IntVal({1}) int uc = (@IntVal({1}) int) input;
+    }
+
     public void IntDoubleTest(
             @IntVal({0, 1}) int iv,
             @IntRange(from = 2, to = 3) int ir,

--- a/framework/tests/value/Basics.java
+++ b/framework/tests/value/Basics.java
@@ -169,6 +169,8 @@ class Basics {
         @IntVal({0, 1, 2}) int sc = (@IntVal({0, 1, 2}) int) input;
         //:: warning: (cast.unsafe)
         @IntVal({1}) int uc = (@IntVal({1}) int) input;
+        //:: warning: (cast.unsafe)
+        @IntVal({2}) int bc = (@IntVal({2}) int) input;
     }
 
     public void IntDoubleTest(

--- a/framework/tests/value/Basics.java
+++ b/framework/tests/value/Basics.java
@@ -166,6 +166,7 @@ class Basics {
     public void intCastTest(@IntVal({0, 1}) int input) {
         @IntVal({0, 1}) int c = (int) input;
         @IntVal({0, 1}) int ac = (@IntVal({0, 1}) int) input;
+        @IntVal({0, 1, 2}) int sc = (@IntVal({0, 1, 2}) int) input;
         //:: warning: (cast.unsafe)
         @IntVal({1}) int uc = (@IntVal({1}) int) input;
     }

